src/javascript/url-shortener.js

```diff
@@ -12,71 +12,327 @@
+  const response = await fetch(url, options);
+
+  if (checkForbidden && response.status === 403) {
+    throw forbiddenError;
+  }
+
+  const body = await response.json();
+
+  if (!response.ok) {
+    throw new Error(`Error ${response.code} from ${url}: ${body.message}`);
+  }
+
+  return body;
+};
 
 
 /**
- * Returns the cached API load promise or calls `loadUrlShortenerApi` and
- * caches the result.
- * @return {Promise} A promise resolved once the library is loaded.
+ * Attempt to create a short URL by POSTing to the bitly API. Return the
+ * created short link. The bitly create-link API requires a group, and doesn't
+ * have any notion of a default group, so we also look up a group to use.
  */
-function urlShortenerApiReady() {
-  return urlShortenerLoadPromise ||
-      (urlShortenerLoadPromise = loadUrlShortenerApi());
-}
+const createBitlink = ({longUrl, token, checkForbidden}) =>
+  getBitlyGroup({token, checkForbidden})
+  .then(guid => createBitlinkCall({longUrl, token, checkForbidden, guid}));
 
 
 /**
- * Accepts a long URL and returns a promise that is resolved with its shortened
- * version. This function loads the urlshortener API if it's not loaded, but
- * the library is only ever loaded once. Long URLs are also cached, so multiple
- * requests for the same URL are never made.
- * @param {string} longUrl
- * @return {Promise} A promise resolved with the shortend URL.
+ * Create a bitlink with the bitly API.
  */
-export function shortenUrl(longUrl) {
-  return new Promise((resolve, reject) => {
-    if (urlMapCache[longUrl]) {
-      resolve(urlMapCache[longUrl]);
-    } else {
-      urlShortenerApiReady().then((urlshortener) => {
-        urlshortener.insert({longUrl}).then(({result}) => {
-          resolve(result.id);
-        }, ({result}) => {
-          reject(new Error(result.error.message));
-        });
-      });
-    }
+const createBitlinkCall = ({longUrl, token, guid, checkForbidden}) =>
+  bitlyApiFetch({
+    token: token,
+    endpoint: '/shorten',
+    checkForbidden: checkForbidden,
+    payload: {
+      long_url: longUrl,
+      group_guid: guid,
+    },
+  }).then(data => data.link);
+
+
+/**
+ * Get the user's "default" bitly group. This group is cached as in localStorage
+ * using BITLY_GUID_STORAGE_KEY, and looked up with that key. If the key doesn't
+ * exist, use the bitly API to fetch the group.
+ */
+const getBitlyGroup = ({token, checkForbidden}) => {
+  let guid = localStorage.getItem(BITLY_GUID_STORAGE_KEY);
+  return guid ?
+    Promise.resolve(guid) :
+    fetchBitlyGroup({token, checkForbidden}).then(guid => {
+      localStorage.setItem(BITLY_GUID_STORAGE_KEY, guid);
+      return guid;
+    });
+};
+
+// Get the user's bitly group. Users can have more than one group, and
+// we don't really have a good way to identify which one is the "primary",
+// so for now we just return the first one in the array.
+// TODO(nathanwest): that's almost certainly not the behavior we actually
```

Comment:
Does this need to be fixed before merging?

Reply:
I wasn't going to worry too much about it, since I'm assuming most people won't be using a "professional" bit.ly account. bit.ly doesn't document what a "group" even is, so I'm not sure what the correct behavior is. I can change it to reject if people have more than 1 group, with a message saying to please reach out so we can decide the correct behavior. I'm not interested in adding any extra interface for users to select a group; this interface is supposed to be a simple "click button" -> "shorten link", just like you'd get on bitly's home page.